### PR TITLE
#9151: Add optional tensor to embeddings

### PR DESCRIPTION
**Pull request by Aswinmcw**
Add optional output tensor and queue id for embeddings forward and backward op

All Post commit test : https://github.com/tenstorrent/tt-metal/actions/runs/9415786167

### CODE REVIEW

File: tt_eager/tt_dnn/op_library/embeddings/embeddings_op.cpp [hunk 1]
```diff
@@ -509,17 +509,23 @@ operation::ProgramWithCallbacks embeddings_(
...
+        const auto& out_tensor = output_tensors.at(0).value();
+        TT_FATAL(out_tensor.get_legacy_shape() == output_shape_required.at(0), fmt::format("The input tensors need a shape of {}, however the output tensor is only {}", output_shape_required,  out_tensor.get_legacy_shape()));
+    }
     TT_FATAL(a.get_dtype() == DataType::UINT32 or a.get_dtype() == DataType::BFLOAT16, "Input must be UINT32 or BFLOAT16");
     TT_FATAL(weights.get_dtype() == DataType::BFLOAT16);
     TT_FATAL(a.memory_config().memory_layout == TensorMemoryLayout::INTERLEAVED, "Embedding does not currently support sharding");
     TT_FATAL(weights.memory_config().memory_layout == TensorMemoryLayout::INTERLEAVED, "Embedding does not currently support sharding");
-    TT_FATAL(this->output_mem_config.memory_layout == TensorMemoryLayout::INTERLEAVED, "Embedding does not currently support sharding");
+    auto out_mem_config = (!output_tensors.empty() && output_tensors.at(0).has_value()) ? output_tensors.at(0).value().memory_config() : this->output_mem_config;
+    TT_FATAL(out_mem_config.memory_layout == TensorMemoryLayout::INTERLEAVED, "Embedding does not currently support sharding");
```
Comment (tt-aho) [inline tt_eager/tt_dnn/op_library/embeddings/embeddings_op.cpp:528]: There is an assertion on output dtype param after this that needs to be changed to the output tensor's dtype when output tensor is passed in

**Review by tt-aho — CHANGES_REQUESTED** [review]

File: tt_eager/tt_dnn/op_library/embeddings/embeddings_op.cpp (hunk 1 above)
Comment (Aswinmcw) [reply tt_eager/tt_dnn/op_library/embeddings/embeddings_op.cpp:531]: Modified the dtype based on output tensor

**Review by TT-BrianLiu — APPROVED** [review]

File: tt_eager/tt_lib/csrc/tt_lib_bindings_tensor.cpp [hunk 2]
```diff
@@ -674,14 +674,15 @@ void TensorModule(py::module& m_tensor) {
     // input embeddings
     m_tensor.def(
         "embeddings",
-        &embeddings,
+        py::overload_cast<const Tensor&, const Tensor&, bool, EmbeddingsType, std::optional<uint32_t>, const MemoryConfig&, std::optional<const DataType>, std::optional<Tensor> >(&embeddings),
```
Comment (tt-aho) [inline tt_eager/tt_lib/csrc/tt_lib_bindings_tensor.cpp:677]: We don't need 2 bindings since queue id is an optional arg

**Review by tt-aho — CHANGES_REQUESTED** [review]

**Review by eyonland — APPROVED** [review]

File: tt_eager/tt_lib/csrc/tt_lib_bindings_tensor.cpp (hunk 2 above)
Comment (Aswinmcw) [reply tt_eager/tt_lib/csrc/tt_lib_bindings_tensor.cpp]: I have removed one bindings, Kindly review it once

**Comment by Aswinmcw** [comment]
Closing as optional tensor support and queue id is already provided to embeddings

**Pull request closed without merging**